### PR TITLE
don't instantiate instance if not needed in sensor execution

### PR DESCRIPTION
**Pull request by neilfulwiler**
## Summary & Motivation

We only use the CachingDynamicPartitionsLoader when there's a run request with partitions, and the CachingDynamicPartitionsLoader needs to hydrate the instance ref which is costly. to speed up the typical case, use @functools.cache to memoize the creation of the `CachingDynamicPartitionsLoader(context.instance)`

## How I Tested These Changes

Ran 10k no-op sensor executions before and after the change, this dropped the latency from 20s to 2s

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/sensor_definition.py [hunk 1]
```diff
@@ -1030,9 +1031,11 @@ def _get_repo_job_by_name(context: SensorEvaluationContext, job_name: str) -> Jo
...
             ]
 
-        dynamic_partitions_store = (
-            CachingDynamicPartitionsLoader(context.instance) if context.instance_ref else None
-        )
+        @functools.cache
+        def dynamic_partitions_store():
+            return (
+                CachingDynamicPartitionsLoader(context.instance) if context.instance_ref else None
+            )
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/_core/definitions/sensor_definition.py:1038]: `@functools.cache` is 3.9+ and we haven't quite dropped 3.8 support yet 

I think a memoized getter (`@cached_property` or manual) on `SensorEvaluationContext` is a bit more straight forward than this cached inline function

Comment (neilfulwiler) [reply python_modules/dagster/dagster/_core/definitions/sensor_definition.py:1038]: ugh i keep forgetting about that. ok thanks

**Review by alangenfeld — APPROVED** [review]
nice

**Comment by neilfulwiler** [comment]
### Merge activity

* **Oct 22, 1:41 PM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/25429).

**Pull request merged**